### PR TITLE
Wip

### PR DESCRIPTION
**Pull request by Rohanksingh18**
Added Some more Tests cases

### CODE REVIEW

File: ecom_test/src/api_helpers/CouponsAPIHelper.py [hunk 1]
```diff
@@ -22,7 +22,7 @@ def call_create_coupon(self, payload):
     def call_retrieve_coupon(self, coupon_id):
         """
         Calling 'retrieve_coupon'the get method of
-        the woo_api_utility instance with the coupon ID as a parameter.
+        the woo_api_utility instance with the Coupon_ID as a parameter.
```
Comment (admaskinfu) [inline ecom_test/src/api_helpers/CouponsAPIHelper.py]: coupon_id or Coupon_ID?

File: ecom_test/src/generic_helpers/generic_coupon_helper.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+            expiration_date = datetime.now().isoformat()
+        else:
+            expiration_date = None
+
+        if not coupon_code:
+            coupon_code = ''.join(random.choice(string.ascii_uppercase) for i in range(length))
+
+        payload = {
+            "code": coupon_code,
+            "discount_type": "percent",
```
Comment (admaskinfu) [inline ecom_test/src/generic_helpers/generic_coupon_helper.py]: You have a parameter for 'discount_type' so you should not hard code "percent" here. use the parameter.
Same for 'amount'.

File: ecom_test/tests/backend_tests/product_reviews/test_prodcuts_review_smoke.py [hunk 3]
```diff
@@ -74,7 +74,9 @@ def test_verify_review_creation_fails_when_fields_duplicated(self):
 
     @pytest.mark.tcid50
     def test_review_creation_should_not_fail_if_reviewer_is_different(self):
-        """All fields have the same value but the 'reviewer' is changes. Expecting successful review creation."""
+        """All fields have the same value but the 'reviewer' is changes.
+           Expecting successful review creation.
+        """
         # Create a review for the product
         review_rs = self.product_review_helper.create_random_review_for_product(self.product_id)
```
Comment (admaskinfu) [inline ecom_test/tests/backend_tests/product_reviews/test_prodcuts_review_smoke.py:82]: this is interesting. we called the 'crate_random_review...' function but we didn't check if response is correct.

File: ecom_test/tests/backend_tests/product_reviews/test_products_reviews_statuses.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    # Create instances of the necessary helper classes
+    request.cls.product_helper = GenericProductHelper()
+    request.cls.product_review_helper = GenericProductReviewsHelper()
+
+    # Create a product and store its ID for reference
+    request.cls.product = request.cls.product_helper.create_a_product()
+    request.cls.rating_count_before = request.cls.product['rating_count']
+    request.cls.product_id = request.cls.product['id']
+
+    yield request.cls.product_helper, request.cls.product_review_helper, request.cls.product_id
```
Comment (admaskinfu) [inline ecom_test/tests/backend_tests/product_reviews/test_products_reviews_statuses.py:19]: I dont think you have to yield all those objects. you can just say 'yield'. You are setting the variables as request.cls so they are available as self. in the tests that use the fixture.
I think so. Try it and let me know if i am wrong.

File: ecom_test/tests/backend_tests/product_reviews/test_products_reviews_statuses.py [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    # They become available to the test function that consumes the fixture
+
+    # Perform cleanup by deleting the product created for the test
+    logger.info(f"Running teardown. Deleting product id: {request.cls.product_id}")
+    request.cls.product_helper.delete_a_product(request.cls.product_id)
```
Comment (admaskinfu) [inline ecom_test/tests/backend_tests/product_reviews/test_products_reviews_statuses.py:24]: nice cleanup

File: ecom_test/tests/backend_tests/product_reviews/test_products_reviews_statuses.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+
+    @pytest.mark.status
+    @pytest.mark.parametrize("review_status", [
+        pytest.param('hold', marks=[pytest.mark.tcid54]),  # Test case ID: 54, review status: hold
+        pytest.param('spam', marks=[pytest.mark.tcid55]),  # Test case ID: 55, review status: spam
+        pytest.param('unspam', marks=[pytest.mark.tcid56]),  # Test case ID: 56, review status: un-spam
+        pytest.param('trash', marks=[pytest.mark.tcid57]),  # Test case ID: 57, review status: trash
+        pytest.param('untrash', marks=[pytest.mark.tcid58])  # Test case ID: 58, review status: un-trash
+    ])
+    def test_verify_create_review_with_different_statuses(self, setup_teardown, review_status, rating=5):
```
Comment (admaskinfu) [inline ecom_test/tests/backend_tests/product_reviews/test_products_reviews_statuses.py:42]: do you have pass 'setup_teardown' even if you are using @pytest.mark.usefixtures? I am not sure. can you try without passing in setup_teardown fixture as a parameter?

File: ecom_test/tests/backend_tests/product_reviews/test_products_reviews_statuses.py [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+        expected_status = 'hold' if review_status in ('un-trash', 'un-spam') else review_status
+
+        # Verify that the created review has the expected status
+        assert review_retrieved['status'] == expected_status, \
+            f"Failed to create review with status = '{expected_status}'."
+
+    @pytest.mark.tcid59
+    def test_verify_review_can_not_be_created_with_invalid_status(self):
+        """
+        Test to verify that a review cannot be created with an invalid status.
```
Comment (admaskinfu) [inline ecom_test/tests/backend_tests/product_reviews/test_products_reviews_statuses.py:89]: nice doc string

File: ecom_test/tests/frontend_tests/cart/test_verify_expired_coupon_message.py [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+    @pytest.mark.tcid11
+    def test_expired_coupon_message(self, setup):
+        """
+        Test case to verify that an expired coupon displays the correct error message.
+
+        Args:
+            setup: Fixture to set up the test environment.
+        """
+        # Step 1: Go to the home page
```
Comment (admaskinfu) [inline ecom_test/tests/frontend_tests/cart/test_verify_expired_coupon_message.py:49]: nice test. nice comments.

**Review by admaskinfu — COMMENTED** [review]
I think there is like 1 comment you actually have to change before merging. the rest are just random random thing to think about.

File: ecom_test/src/api_helpers/CouponsAPIHelper.py (hunk 1 above)
Comment (Rohanksingh18) [reply ecom_test/src/api_helpers/CouponsAPIHelper.py]: Indeed

File: ecom_test/src/generic_helpers/generic_coupon_helper.py (hunk 2 above)
Comment (Rohanksingh18) [reply ecom_test/src/generic_helpers/generic_coupon_helper.py]: Yes, you are right. I changed

File: ecom_test/tests/backend_tests/product_reviews/test_prodcuts_review_smoke.py (hunk 3 above)
Comment (Rohanksingh18) [reply ecom_test/tests/backend_tests/product_reviews/test_prodcuts_review_smoke.py:82]: Yes, I need to add more assertions functions to check the response for a few more tests not only for this test only. I will rectify this issue

Comment (Rohanksingh18) [reply ecom_test/tests/backend_tests/product_reviews/test_prodcuts_review_smoke.py:82]: I did not test using Postman, which is why I did not add response functions now I added them to verify the response.

File: ecom_test/tests/backend_tests/product_reviews/test_products_reviews_statuses.py (hunk 4 above)
Comment (Rohanksingh18) [reply ecom_test/tests/backend_tests/product_reviews/test_products_reviews_statuses.py:19]: Ya, this time I wrote differently as I did not use ".self" to call the function in tests. And it is true that I set the variables as request class; however, I did not call any function by calling  .self. But I will take care of this in future

File: ecom_test/tests/backend_tests/product_reviews/test_products_reviews_statuses.py (hunk 5 above)
Comment (Rohanksingh18) [reply ecom_test/tests/backend_tests/product_reviews/test_products_reviews_statuses.py:24]: Thank you

File: ecom_test/tests/backend_tests/product_reviews/test_products_reviews_statuses.py (hunk 6 above)
Comment (Rohanksingh18) [reply ecom_test/tests/backend_tests/product_reviews/test_products_reviews_statuses.py:42]: Yes, I tried initially but it did not work well and also did not give auto suggestions for the function (when I call). By adding setup_teardown I can track any function. 
image 
without setup_teardown: 
image

**Pull request merged**